File: packages/core-js/src/convert.ts
import type { JsonmdOptions, JsonValue } from './types.js';
import { toTable, toKeyValueTable, toBulletList, isPrimitive, isFlat } from './table.js';
import { isMcpToolList, mcpToMd } from './mcp.js';

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function renderNested(data: unknown, depth: number, maxDepth: number): string {
  if (depth > maxDepth) {
    return '`' + JSON.stringify(data) + '`';
  }

  if (data === null || data === undefined) return '';
  if (isPrimitive(data)) return String(data);

  if (Array.isArray(data)) {
    if (data.length === 0) return '';

    // Array of primitives
    if (data.every(isPrimitive)) {
      return toBulletList(data);
    }

    // Array of flat objects
    if (data.every((item) => isObject(item) && isFlat(item as Record<string, unknown>))) {
      return toTable(data as Record<string, unknown>[]);
    }

    // Array of objects (possibly with nested values) - still try table if mostly flat
    if (data.every(isObject)) {
      const flatCount = data.filter((item) => isFlat(item as Record<string, unknown>)).length;
      if (flatCount / data.length >= 0.8) {
        return toTable(data as Record<string, unknown>[]);
      }
    }

    // Mixed/nested array - render indexed sections
    const sections: string[] = [];
    const heading = '#'.repeat(Math.min(depth + 1, 6));
    for (let i = 0; i < data.length; i++) {
      sections.push(`${heading} [${i}]\n\n${renderNested(data[i], depth + 1, maxDepth)}`);
    }
    return sections.join('\n\n');
  }

  if (isObject(data)) {
    const obj = data as Record<string, unknown>;
    const entries = Object.entries(obj);
    if (entries.length === 0) return '';

    // Flat object
    if (isFlat(obj)) {
      return toKeyValueTable(obj);
    }

    // Nested object - headings + recursion
    const sections: string[] = [];
    const heading = '#'.repeat(Math.min(depth + 1, 6));
    for (const [key, value] of entries) {
      if (isPrimitive(value)) {
        sections.push(`${heading} ${key}\n\n${value === null ? '' : String(value)}`);
      } else {
        sections.push(`${heading} ${key}\n\n${renderNested(value, depth + 1, maxDepth)}`);
      }
    }
    return sections.join('\n\n');
  }

  return String(data);
}

/**
 * Smart convert JSON data to Markdown, auto-detecting the best representation.
 */
export function smartConvert(data: unknown, opts?: JsonmdOptions): string {
  const maxDepth = opts?.maxDepth ?? 6;
  const mode = opts?.mode ?? 'table';

  // Check MCP first
  if (isMcpToolList(data)) {
    return mcpToMd(data, mode);
  }

  const parts: string[] = [];

  if (opts?.title) {
    parts.push(`# ${opts.title}`);
    parts.push('');
  }

  const body = renderNested(data, 1, maxDepth);
  if (body) parts.push(body);

  return parts.join('\n');
}

/**
 * Alias for smartConvert with title support.
 */
export function jsonToMd(data: unknown, title?: string): string {
  return smartConvert(data, title ? { title } : undefined);
}

File: packages/core-python/jsonmd/convert.py
"""Core JSON to Markdown conversion."""

from __future__ import annotations

import json
from typing import Any

from .table import to_table, to_key_value_table, to_bullet_list, is_primitive, is_flat
from .mcp import is_mcp_tool_list, mcp_to_md


def _is_object(value: Any) -> bool:
    return isinstance(value, dict)


def _render_nested(data: Any, depth: int, max_depth: int) -> str:
    if depth > max_depth:
        return "`" + json.dumps(data) + "`"

    if data is None:
        return ""
    if is_primitive(data):
        if isinstance(data, bool):
            return str(data).lower()
        return str(data)

    if isinstance(data, list):
        if len(data) == 0:
            return ""

        # Array of primitives
        if all(is_primitive(item) for item in data):
            return to_bullet_list(data)

        # Array of flat objects
        if all(_is_object(item) and is_flat(item) for item in data):
            return to_table(data)

        # Array of objects (possibly with nested) - try table if mostly flat
        if all(_is_object(item) for item in data):
            flat_count = sum(1 for item in data if is_flat(item))
            if flat_count / len(data) >= 0.8:
                return to_table(data)

        # Mixed/nested array
        heading = "#" * min(depth + 1, 6)
        sections = []
        for i, item in enumerate(data):
            sections.append(f"{heading} [{i}]\n\n{_render_nested(item, depth + 1, max_depth)}")
        return "\n\n".join(sections)

    if _is_object(data):
        if len(data) == 0:
            return ""

        if is_flat(data):
            return to_key_value_table(data)

        # Nested object
        heading = "#" * min(depth + 1, 6)
        sections = []
        for key, value in data.items():
            if is_primitive(value):
                display = "" if value is None else (str(value).lower() if isinstance(value, bool) else str(value))
                sections.append(f"{heading} {key}\n\n{display}")
            else:
                sections.append(f"{heading} {key}\n\n{_render_nested(value, depth + 1, max_depth)}")
        return "\n\n".join(sections)

    return str(data)


def smart_convert(data: Any, *, mode: str = "table", max_depth: int = 6, title: str | None = None) -> str:
    """Smart convert JSON data to Markdown, auto-detecting the best representation."""
    if is_mcp_tool_list(data):
        return mcp_to_md(data, mode)

    parts = []
    if title:
        parts.append(f"# {title}")
        parts.append("")

    body = _render_nested(data, 1, max_depth)
    if body:
        parts.append(body)

    return "\n".join(parts)


def json_to_md(data: Any, title: str | None = None) -> str:
    """Alias for smart_convert with title support."""
    return smart_convert(data, title=title)
